File: go/ManifoldApi/GetMarketPositions.go
package ManifoldApi

import (
	"encoding/json"
	"fmt"
)

type MarketPosition []struct {
	From struct {
		Day struct {
			Value         float64 `json:"value"`
			Profit        int     `json:"profit"`
			Invested      float64 `json:"invested"`
			PrevValue     float64 `json:"prevValue"`
			ProfitPercent int     `json:"profitPercent"`
		} `json:"day"`
		Week struct {
			Value         float64 `json:"value"`
			Profit        int     `json:"profit"`
			Invested      float64 `json:"invested"`
			PrevValue     float64 `json:"prevValue"`
			ProfitPercent int     `json:"profitPercent"`
		} `json:"week"`
		Month struct {
			Value         float64 `json:"value"`
			Profit        float64 `json:"profit"`
			Invested      float64 `json:"invested"`
			PrevValue     float64 `json:"prevValue"`
			ProfitPercent float64 `json:"profitPercent"`
		} `json:"month"`
	} `json:"from"`
	Loan        float64 `json:"loan"`
	Payout      float64 `json:"payout"`
	Profit      float64 `json:"profit"`
	UserID      string  `json:"userId"`
	Invested    float64 `json:"invested"`
	UserName    string  `json:"userName"`
	HasShares   bool    `json:"hasShares"`
	ContractID  string  `json:"contractId"`
	HasNoShares bool    `json:"hasNoShares"`
	LastBetTime int64   `json:"lastBetTime"`
	TotalShares struct {
		NO  int     `json:"NO"`
		YES float64 `json:"YES"`
	} `json:"totalShares"`
	HasYesShares     bool    `json:"hasYesShares"`
	UserUsername     string  `json:"userUsername"`
	ProfitPercent    float64 `json:"profitPercent"`
	UserAvatarURL    string  `json:"userAvatarUrl"`
	MaxSharesOutcome string  `json:"maxSharesOutcome"`
}

func GetMarketPositions(marketId string) []MarketPosition {
	sb := callManifoldApi("GET", fmt.Sprintf("v0/market/%s/positions", marketId), nil)

	var response []MarketPosition
	json.Unmarshal([]byte(sb), &response)
	return response
}

File: ts/src/api/getMarketPositions.ts
import { callManifoldApi } from "./callManifoldApi.ts";
import { z } from "zod";

export const ContractMetricSchema = z.object({
  contractId: z.string(),
  answerId: z.string().nullable(),

  from: z
    .record(
      z.string(),
      z.object({
        profit: z.number(),
        profitPercent: z.number(),
        invested: z.number(),
        prevValue: z.number(),
        value: z.number(),
      }),
    )
    .optional(),

  hasNoShares: z.boolean(),
  hasShares: z.boolean(),
  hasYesShares: z.boolean(),

  invested: z.number(),
  loan: z.number(),

  maxSharesOutcome: z.string().nullable(),

  payout: z.number(),
  profit: z.number(),
  profitPercent: z.number(),

  totalShares: z.record(z.string(), z.number()),

  userId: z.string(),
  userUsername: z.string().optional(),
  userName: z.string().optional(),
  userAvatarUrl: z.string().optional(),

  lastBetTime: z.number(),
});

export type MarketPosition = z.infer<typeof ContractMetricSchema>;

export async function getMarketPositionsForUser(
  marketId: string,
  userId: string,
): Promise<MarketPosition[]> {
  const sb = await callManifoldApi(
    "GET",
    `v0/market/${marketId}/positions`,
    {
      userId,
    },
  );
  return z.array(ContractMetricSchema).parse(JSON.parse(sb));
}
